### PR TITLE
Producer SpanKind does not always imply a remote call

### PR DESCRIPTION
**Pull request by lmolkova**
It's popular in asynchronous scenarios to send messages in batches.

In this case, each message should carry a unique context to be individually traceable. Arguably it means creating span per message. Such spans do not represent remote calls.

Sending a batch (which is remote outgoing call) is represented by a different span that is linked to all of the messages span contexts.

So async messaging scenarios with batching involve
* *message* spans with PRODUCER kind (each of them is a parent of asynchronous request)
* *send* span that is CLIENT kind (which is remote outgoing call). But it is not a parent of any asynchronous request

In some cases (not involving batching) it is possible that dedicated *message* span is not needed and *send* may represent both: context in the message and remote outgoing call. In this case, such span should have PRODUCER kind

### CODE REVIEW

**Comment by jmacd** [comment]
Could you update the spec language to explain why PRODUCER may or may not be remote? Also I suspect that if PRODUCER has this property, so does CONSUMER. Should they both be marked maybes?

**Comment by lmolkova** [comment]
@jmacd sure, done.

I wish we had a spec for async messaging documented, but it should not block v1.  I will try to come up with one

**Comment by carlosalberto** [comment]
Not sure "may be" is clear enough, but otherwise it looks good to me.

File: specification/api-tracing.md [hunk 1]
```diff
@@ -577,9 +577,13 @@ These are the possible SpanKinds:
 * `PRODUCER` Indicates that the span describes the parent of an
   asynchronous request.  This parent span is expected to end before
   the corresponding child `CONSUMER` span, possibly even before the
-  child span starts.
+  child span starts. In messaging scenarios with batching, tracing
+  individual messages requires a new span per message to be created (with
+  `PRODUCER` kind), while publishing a batch is represented by another span
```
Comment (jmacd) [inline specification/api-tracing.md]: Strictly speaking, I suppose you could publish a batch using an INTERNAL span too, I think. Would you object to just removing the `, while publishing ... (with CLIENT kind)` part of this sentence?

File: specification/api-tracing.md [hunk 2]
```diff
@@ -577,9 +577,13 @@ These are the possible SpanKinds:
...
+  (with `CLIENT` kind).
 * `CONSUMER` Indicates that the span describes the child of an
-  asynchronous remote `PRODUCER` request.
+  asynchronous `PRODUCER` request. Asynchronous scenarios may
```
Comment (jmacd) [inline specification/api-tracing.md]: I think the `Asynchronous scenarios ...` sentence should move up to the paragraph around line 560 where asynchronous vs synchronous is discussed.

File: specification/api-tracing.md [hunk 3]
```diff
@@ -590,6 +594,6 @@ To summarize the interpretation of these kinds:
 |--|--|--|--|--|
 | `CLIENT` | yes | | | yes |
 | `SERVER` | yes | | yes | |
-| `PRODUCER` | | yes | | yes|
-| `CONSUMER` | | yes | yes | |
+| `PRODUCER` | | yes | | may be |
```
Comment (jmacd) [inline specification/api-tracing.md]: nit: Use "maybe", not "may be" (https://www.grammar.com/maybe_vs._may_be)

File: specification/api-tracing.md [hunk 4]
```diff
@@ -579,9 +579,11 @@ These are the possible SpanKinds:
 * `PRODUCER` Indicates that the span describes the parent of an
   asynchronous request.  This parent span is expected to end before
   the corresponding child `CONSUMER` span, possibly even before the
-  child span starts.
+  child span starts. In messaging scenarios with batching, tracing
```
Comment (Oberon00) [inline specification/api-tracing.md]: How does the producer know if the consumer uses batching?

Also, I think this belongs to semantic conventions.

Comment (SergeyKanzhelev) [reply specification/api-tracing.md]: this note simply expands the scope of producer and consumer types. I agree that the complete solution on how to do messaging may belong to some other document

**Comment by SergeyKanzhelev** [comment]
Active for more than 3 days and have enough sign offs. Merging

**Pull request merged**